gk/fib.c

```diff
@@ -1361,45 +1361,121 @@ check_gateway_prefix(struct ip_prefix *prefix, struct ipaddr *gw_addr)
+ * in deployments of Gatekeeper until it is too late.
  */
 static int
 check_prefix_locked(struct ip_prefix *prefix,
 	enum gk_fib_action action, struct gk_config *gk_conf)
 {
 	uint8_t i;
 	int ip_prefix_present;
+	uint32_t prefix_ip4;
 	struct gk_fib *ip_prefix_fib;
 	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
 
-	if (action == GK_DROP || action == GK_FWD_GRANTOR)
+	if (action == GK_DROP || action == GK_FWD_GRANTOR) {
+		/* Ensure that all prefixes longer than @prefix are safe. */
+		if (prefix->addr.proto == ETHER_TYPE_IPv4) {
+			uint32_t next = 0;
+			int ret;
+			struct rte_lpm_iterator_state state;
+			const struct rte_lpm_rule *re;
+
+			prefix_ip4 = ntohl(prefix->addr.ip.v4.s_addr);
+			ret = rte_lpm_iterator_state_init(
+				prefix_ip4, prefix->len, &state);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to initialize the lpm rule iterator state at %s!\n",
+					__func__);
+				return -1;
+			}
+
+			ret = rte_lpm_rule_iterate(
+				ltbl->lpm, &state, &next, &re);
+			while (ret >= 0) {
+				ip_prefix_fib = &ltbl->fib_tbl[re->next_hop];
+				if (ip_prefix_fib->action != GK_FWD_GRANTOR &&
+						ip_prefix_fib->action !=
+						GK_DROP) {
+					RTE_LOG(WARNING, GATEKEEPER,
+						"gk: adding this rule with prefix %s and action %u would add a security hole since there already exists an entry of %u length with action %u",
+						prefix->str, action, state.depth,
+						ip_prefix_fib->action);
+					return -1;
+				}
+				ret = rte_lpm_rule_iterate(
+					ltbl->lpm, &state, &next, &re);
+			}
+		} else if (likely(prefix->addr.proto == ETHER_TYPE_IPv6)) {
+			uint32_t next = 0;
+			struct rte_lpm6_iterator_state state;
+			const struct rte_lpm6_rule *re;
+			int ret = rte_lpm6_iterator_state_init(
+				prefix->addr.ip.v6.s6_addr,
+				prefix->len, &state);
+			if (ret) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to initialize the lpm6 rule iterator state at %s!\n",
+					__func__);
+				return -1;
+			}
+
+			ret = rte_lpm6_rule_iterate(
+				ltbl->lpm6, &state, &next, &re);
+			while (ret >= 0) {
+				ip_prefix_fib = &ltbl->fib_tbl6[re->next_hop];
+				if (ip_prefix_fib->action != GK_FWD_GRANTOR &&
+						ip_prefix_fib->action !=
+						GK_DROP) {
+					RTE_LOG(WARNING, GATEKEEPER,
+						"gk: adding this rule with prefix %s and action %u would add a security hole since there already exists an entry of %u length with action %u",
+						prefix->str, action, re->depth,
+						ip_prefix_fib->action);
+					return -1;
+				}
+				ret = rte_lpm6_rule_iterate(
+					ltbl->lpm6, &state, &next, &re);
+			}
+		} else {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: unknown IP type %hu with prefix %s and action %u",
+				prefix->addr.proto, prefix->str, action);
+			return -1;
+		}
+
 		return 0;
+	}
 
+	/* Ensure that the new prefix does not create a security hole. */
 	if (prefix->addr.proto == ETHER_TYPE_IPv4) {
+		prefix_ip4 = ntohl(prefix->addr.ip.v4.s_addr);
```

Comment:
Since variable `prefix_ip4` is only needed here now, move its declaration here as well.